### ISSUE TITLE
Paint tool in segmentation editor doesn't paint mouse trajectory area continuously

### ISSUE THREAD

**Issue by goldragoon** [Slicer/Slicer#5057]
## Steps to reproduce

Whenever I moving mouse too fast, paint tool paints too sparsely along mouse trajectory as below.
![ezgif com-crop](https://user-images.githubusercontent.com/2537598/88404898-7a4b3c80-ce09-11ea-96e8-f5d897c2e38a.gif)

## Expected behavior

I think that paint functionality should fill area along mouse trajectory continuously.
If this is not intended, I'd like to fix it by myself.

## Environment
- Slicer version: Slicer-4.10.2
- Operating system: Windows

**Comment by lassoan** [comment]
Thank you for reporting this. For this to be an issue, you need to move the mouse really fast, have a very slow computer, or have very slow rendering (e.g., CPU volume rendering) in the 3D view.

Anyway, if you want to give it a try and add interpolated points between distant mouse positions then we would be certainly interested in trying it. If it helps in some scenarios and does not impact normal use cases then it should be no problem to integrate it.

**Comment by goldragoon** [comment]
Thanks @lassoan. I'll try it.

**Comment by fusentasticus** [comment]
Hi, my vote is that @goldragoon found a regression. 

I have compared 4.11.0-2020-07-20 (recent) with 4.11.0.2020-01-08 (just pre-COVID fwiiw). I have disabled, but not uninstalled extensions. I'm running Windows 10. 

My tablet (Wacom) is unusable in recent version with Paint & Erase: hovering is tracked OK, but as soon as the tip touches the surface, the yellow circle freezes. The circle gets unstuck only after many pixels of movement of pen on surface. This leaves a path that is broken at its beginning. (Yes, I updated Wacom drivers and checked that pen works in other programs!!!)  

The problem also exists for mouse use, especially if larger volumes are used.  Then a path will become broken up as reported. 

Also, the scroll bar of the Module Panel strangely flickers when mousing around in a viewer.

**Comment by lassoan** [comment]
This issue is about simulating more clicks when events do not arrive frequently enough. Pens and digitizing tablets have the opposite problem - they generate too many events (at least, more than we would need and want to process for segmentation). Please create a separate issue for any pen-related problems that you find. I think we also added pen mode and various configuration options for fine-tuning pen behavior. We'll discuss all these in that new issue. Thank you.

**Comment by fusentasticus** [comment]
@lassoan lassoan Thanks for helpful reply. I just filed #5063 about this specific regression.  I see what you're saying about pen v. mouse events, but I can't really disentangle the two -- both are not currently good.

**Comment by jamesobutler** [comment]
I have also observed the undersampling issue as detailed in the original post.  Specifically if I put MRHead.nrrd into a Red slice only view and use "Paint" to draw a big "X" on screen using the same mouse speed as if I was told to point to the four corners I can observe that behavior.  I'm using a 2019 Lenovo Thinkpad mobile Workstation, but always attributed this type of performance with "Paint" and "Erase" assuming it required extra processing to draw all the little circles on screen and possibly computing as-you-go the voxels that will be filled in.  

@lassoan You probably know more about these effects - Does "Paint" and "Erase" compute more things during the click and hold + mouse move event compared to "Draw" and "Scissors" which appears to do all the processing at the end when you call to close the loop/end the interaction? The latter effects draw quickly without creating straight lines between dropped point locations.  I can quickly draw a circle that looks like a smooth circle.

**Comment by lassoan** [comment]
I have an ultrabook (Surface Laptop 3, with i7) and I don't feel that the update speed would be slow. We don't draw strokes but individual disks or spheres, so there is of course discontinuity, but you would not segment by moving the mouse this quickly anyway. Do you see similar performance as on my computer as shown in this screen capture video:
https://1drv.ms/v/s!Arm_AFxB9yqHw7xBipfIZX_wOcCyxw?e=828hqS

Please use latest Slicer Preview Release (we will not make any changes to Slicer-4.10.2 and we have already made lots of improvements and fixes since Slicer-4.10.2).

If you switch to 4-up view and enable FPS display in the 3D view, what frame rate do you see while painting?

![image](https://user-images.githubusercontent.com/307929/88607938-81af5780-d04e-11ea-9872-e90d237e9e97.png)

What operating system do you use? What graphics card do you have?

> Does "Paint" and "Erase" compute more things during the click and hold + mouse move event compared to "Draw" and "Scissors" which appears to do all the processing at the end when you call to close the loop/end the interaction?

Paint is slower than draw or scissors because the paintbrush is previewed in _all_ the views. This is a very useful feature but slows things down, as it means that all the views are redrawn when the brush is moved. We might be able to do immediate update in the current view and do delayed update on others; or we could make it an application setting to disable preview in other views. We could also interpolate if there is large jump in the mouse coordinates (simulating strokes), which would probably mask this sparse sampling quite effectively.

**Comment by lassoan** [comment]
Interpolation between distant point has been implemented in https://github.com/Slicer/Slicer/commit/52b7208240145c3a6aba99b0a7af0eddaca6f0ff

**Comment by jamesobutler** [comment]
That commit works well @lassoan! Very smooth 👏
